### PR TITLE
[READY]Component Forensics and Item Blood OverlayS!

### PR DESCRIPTION
**Pull request by silicons**
@MoreRobustThanYou's datum forensics but it's a component this time.
Fully refactors forensics, including fingerprints, hidden/admin fingerprints, suit/clothing/whatever fibers, and blood DNA residue to /datum/component/forensics.
Adds support for manually adding fingerprints.
:cl:
code: Forensics is now a datum component.
balance: NPC humans will now start leaving fingerprints on things they touch!
/:cl:

### CODE REVIEW

File: code/datums/components/forensics.dm [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	var/fibertext
+	var/item_multiplier = isitem(src)?1.2:1
+	if(M.wear_suit)
+		fibertext = "Material from \a [M.wear_suit]."
+		if(prob(10*item_multiplier) && !(fibertext in fibers))
+			LAZYADD(fibers, fibertext)
+		if(!(M.wear_suit.body_parts_covered & CHEST))
+			if(M.w_uniform)
+				fibertext = "Fibers from \a [M.w_uniform]."
+				if(prob(12*item_multiplier) && !(fibertext in fibers)) //Wearing a suit means less of the uniform exposed.
```
Comment (RemieRichards) [inline code/datums/components/forensics.dm]: make fibers assoc.

Comment (kevinz000) [reply code/datums/components/forensics.dm]: kk gimme a bit ill do it after i process blood dna (pain)

**Comment by Repukan** [comment]
Cool techwebw brah

**Comment by kevinz000** [comment]
gee i dont suppose i can have that "balance rebalance" changed to a "fix and refactor"? :^)

**Comment by MoreRobustThanYou** [comment]
let's see @kevinz000 break everything again

**Comment by Jordie0608** [comment]
1569 runtimes from one testmerge; come back when you've spent that many hours testing.
https://tgstation13.org/parsed-logs/sybil/data/logs/2017/11/02/round-78254/runtime.txt

File: code/datums/components/forensics.dm [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+		LAZYSET(fingerprints, full_print, full_print)
+
+/datum/component/forensics/proc/add_fibers(list/_fibertext)		//list(text)
+	if(!length(_fibertext))
+		return
+	LAZYINITLIST(fibers)
+	for(var/i in _fibertext)	//We use an associative list, make sure we don't just merge a non-associative list into ours.
+		fibers[i] = i
+
+/datum/component/forensics/proc/add_fibers_from_mob(mob/living/carbon/human/M)
```
Comment (kevinz000) [inline code/datums/components/forensics.dm]: proc runtiming from bad index, putting a note here or me to fix

File: code/datums/components/forensics.dm [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+/datum/component/forensics/proc/wipe_hiddenprints()
+	return	//no.
+
+/datum/component/forensics/proc/wipe_blood_DNA()
+	blood_DNA = null
+
+/datum/component/forensics/proc/wipe_fibers()
+	fibers = null
+
+/datum/component/forensics/proc/add_fingerprints(list/_fingerprints)	//list(text)
```
Comment (Jalleo) [inline code/datums/components/forensics.dm]: Suggestion change this proc name to add_known_fingerprints since this one is all done with copying prints already from a object.

File: code/datums/components/forensics.dm [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	if(!length(_fingerprints))
+		return
+	LAZYINITLIST(fingerprints)
+	for(var/i in _fingerprints)	//We use an associative list, make sure we don't just merge a non-associative list into ours.
+		fingerprints[i] = i
+
+/datum/component/forensics/proc/add_fingerprint_from_mob(mob/living/M, ignoregloves = FALSE)
```
Comment (Jalleo) [inline code/datums/components/forensics.dm]: I would suggest changing this one to only add_fingerprints as per above statement as this one will be typed out a lot.

File: code/datums/components/forensics.dm (hunk 3 above)
Comment (kevinz000) [reply code/datums/components/forensics.dm]: intended

File: code/datums/components/forensics.dm (hunk 4 above)
Comment (kevinz000) [reply code/datums/components/forensics.dm]: intended

**Review by Cyberboss — COMMENTED** [review]
Yelling at myself to review this tomorrow with something about how atom procs kinda defaet the purpose of compenents. Do all these activations have something in common? (attackby, attack_hand, Topic, whatever) they should have signals

**Comment by kevinz000** [comment]
most of the time it's when someone touches something, i could move it to attackby/attack_hand/whatever, but it'd be a pretty powerful buff...

File: code/modules/mob/living/carbon/examine.dm [hunk 5]
```diff
@@ -19,7 +19,7 @@
 
 	for(var/obj/item/I in held_items)
 		if(!(I.flags_1 & ABSTRACT_1))
-			if(I.blood_DNA)
+			if(I.return_blood_DNA())
```
Comment (ShizCalev) [inline code/modules/mob/living/carbon/examine.dm]: Causing #32389

**Comment by kevinz000** [comment]
goddamnit if orgot a length..

**Comment by optimumtact** [comment]
don't you recall our conversation around focus

File: code/datums/components/decals/blood.dm [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+	var/obj/item/I = parent
+	if(initial(I.icon) && initial(I.icon_state))
+		//try to find a pre-processed blood-splatter. otherwise, make a new one
+		var/index = I.blood_splatter_index()
+		var/icon/blood_splatter_icon = GLOB.blood_splatter_icons[index]
+		if(!blood_splatter_icon)
+			blood_splatter_icon = icon(initial(I.icon), initial(I.icon_state), , 1)		//we only want to apply blood-splatters to the initial icon_state for each object
+			blood_splatter_icon.Blend("#fff", ICON_ADD) 			//fills the icon_state with white (except where it's transparent)
+			blood_splatter_icon.Blend(icon('icons/effects/blood.dmi', "itemblood"), ICON_MULTIPLY) //adds blood and the remaining white areas become transparant
+			GLOB.blood_splatter_icons[index] = blood_splatter_icon
```
Comment (ninjanomnom) [inline code/datums/components/decals/blood.dm]: what's the point of making blood a decal when you're going to keep one of the worst examples of old shitcode alive in it

File: code/datums/components/decals/blood.dm [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+		pic = image(blood_splatter_icon, null, initial(I.icon_state), I.layer + 1, null)
```
Comment (ninjanomnom) [inline code/datums/components/decals/blood.dm]: pic has to be a mutable appearance

File: code/datums/components/forensics.dm [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+	return length(fingerprints)? TRUE : FALSE
+
+/datum/component/forensics/proc/has_hiddenprints()
+	return length(hiddenprints)? TRUE : FALSE
+
+/datum/component/forensics/proc/has_blood_DNA()
+	return length(blood_DNA)? TRUE : FALSE
+
+/datum/component/forensics/proc/has_fibers()
+	return length(fibers)? TRUE : FALSE
```
Comment (ninjanomnom) [inline code/datums/components/forensics.dm]: A lovely set of examples of how to not do procs

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/datums/components/forensics.dm (hunk 8 above)
Comment (optimumtact) [reply code/datums/components/forensics.dm]: could use a define though

Comment (optimumtact) [reply code/datums/components/forensics.dm]: HAS_FIBRES length(fibres)

File: code/datums/components/decals/blood.dm (hunk 6 above)
Comment (kevinz000) [reply code/datums/components/decals/blood.dm]: how else do I keep track of generated icons so they aren't regenerated each time?

Comment (ninjanomnom) [reply code/datums/components/decals/blood.dm]: You do realize this was made before overlays were a thing right?

Comment (RemieRichards) [reply code/datums/components/decals/blood.dm]: You do realise this is blending itself onto the icon, such that blood only appears on the parts of the icon that aren't transparent, IE: blood only splatters on the original sprite itself, not blood floating in mid air.

Something overlays can't do.

Comment (ninjanomnom) [reply code/datums/components/decals/blood.dm]: yes but it's creating something that should be applied as an overlay

Comment (RemieRichards) [reply code/datums/components/decals/blood.dm]: which it does, later on, which is now handled in decal component's pic var.
(previously it directly add_overlay()'d) itself.)

Comment (optimumtact) [reply code/datums/components/decals/blood.dm]: new maintainers

**Comment by kevinz000** [comment]
now uses addcomponent and inheritence instead of loadcomponent per ninjanomnom's instructions :3

File: code/datums/components/decals/blood.dm [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+	if(isitem(parent))
+		return ..()
+	else
+		return COMPONENT_INCOMPATIBLE
+
+/datum/component/decal/blood/generate_appearance(_icon, _icon_state, _dir, _layer, _color)
+	var/obj/item/I = parent
+	if(initial(I.icon) && initial(I.icon_state))
+		var/static/list/blood_splatter_icons
+		LAZYINITLIST(blood_splatter_icons)
```
Comment (ninjanomnom) [inline code/datums/components/decals/blood.dm]: just do `var/static/list/blood_splatter_icons = list()`

File: code/datums/components/decals/blood.dm [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+		//try to find a pre-processed blood-splatter. otherwise, make a new one
+		var/index = "[REF(initial(I.icon))]-[initial(I.icon_state)]"
+		var/icon/blood_splatter_icon = blood_splatter_icons[index]
+		if(!blood_splatter_icon)
+			blood_splatter_icon = icon(initial(I.icon), initial(I.icon_state), , 1)		//we only want to apply blood-splatters to the initial icon_state for each object
+			blood_splatter_icon.Blend("#fff", ICON_ADD) 			//fills the icon_state with white (except where it's transparent)
+			blood_splatter_icon.Blend(icon('icons/effects/blood.dmi', "itemblood"), ICON_MULTIPLY) //adds blood and the remaining white areas become transparant
```
Comment (ninjanomnom) [inline code/datums/components/decals/blood.dm]: _icon and _icon_state

File: code/datums/components/forensics.dm [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+/datum/component/forensics/InheritComponent(datum/component/forensics/F, original)
+	fingerprints = F.fingerprints
+	hiddenprints = F.hiddenprints
+	blood_DNA = F.blood_DNA
+	fibers = F.fibers
+	return ..()
+
+/datum/component/forensics/Initialize()
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
```
Comment (ninjanomnom) [inline code/datums/components/forensics.dm]:
```
. = COMPONENT_INCOMPATIBLE
CRASH(message)
```

We want to know when components are applied incorrectly like this

File: code/modules/detectivework/detective_work.dm [hunk 12]
```diff
@@ -1,119 +1,140 @@
...
 
-/atom/proc/transfer_fingerprints_to(atom/A)
+/atom/proc/_wipe_fibers()
+	GET_COMPONENT(D, /datum/component/forensics)
+	return D? D.wipe_fibers() : TRUE
+
+/atom/proc/add_fingerprints(list/fingerprints)
+	if(length(fingerprints))
+		var/datum/component/forensics/D = AddComponent(/datum/component/forensics)
+		. = D.add_fingerprints(fingerprints)
```
Comment (ninjanomnom) [inline code/modules/detectivework/detective_work.dm]: this kind of information should be passed in the add component args

File: code/_onclick/ai.dm [hunk 13]
```diff
@@ -166,15 +166,15 @@
 /obj/machinery/power/apc/AICtrlClick() // turns off/on APCs.
 	if(can_use(usr, 1))
 		toggle_breaker()
-		add_fingerprint(usr)
+		add_fingerprint_from_mob(usr)
```
Comment (ninjanomnom) [inline code/_onclick/ai.dm]: as I said a while back in irc, revert this change and it'll be far less annoying to review this

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/datums/components/decals/blood.dm [hunk 14]
```diff
@@ -0,0 +1,31 @@
+
+/datum/component/decal/blood
+	dupe_mode = COMPONENT_DUPE_UNIQUE
+
+/datum/component/decal/blood/Initialize(_icon, _icon_state, _dir, _cleanable=CLEAN_STRENGTH_BLOOD, _color, _layer=ABOVE_OBJ_LAYER)
+	if(isitem(parent))
+		return ..()
+	else
+		. = COMPONENT_INCOMPATIBLE
+		CRASH("Warning: Blood decal attempted to be added to non-item of type [parent.type]")
```
Comment (ninjanomnom) [inline code/datums/components/decals/blood.dm]: just do if(!isitem) crash

It makes it easier to read if it ever becomes necessary to add more stuff here

File: code/datums/components/decals/blood.dm [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+	var/obj/item/I = parent
+	if(!_icon)
+		_icon = 'icons/effects/blood.dmi'
+	if(!_icon_state)
+		_icon_state = "itemblood"
+	if(initial(I.icon) && initial(I.icon_state))
+		var/static/list/blood_splatter_icons = list()
+		//try to find a pre-processed blood-splatter. otherwise, make a new one
+		var/index = "[REF(initial(I.icon))]-[initial(I.icon_state)]"
+		var/icon/blood_splatter_icon = blood_splatter_icons[index]
```
Comment (ninjanomnom) [inline code/datums/components/decals/blood.dm]: This may as well be a mutable appearance even in the list

File: code/datums/components/decals/blood.dm [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+		. = COMPONENT_INCOMPATIBLE
+		CRASH("Warning: Blood decal attempted to be added to non-item of type [parent.type]")
+
+/datum/component/decal/blood/generate_appearance(_icon, _icon_state, _dir, _layer, _color)
+	var/obj/item/I = parent
+	if(!_icon)
+		_icon = 'icons/effects/blood.dmi'
+	if(!_icon_state)
+		_icon_state = "itemblood"
+	if(initial(I.icon) && initial(I.icon_state))
```
Comment (ninjanomnom) [inline code/datums/components/decals/blood.dm]: If this isn't the case just return false immediately

File: code/datums/components/forensics.dm [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+	var/list/fingerprints		//assoc print = print
+	var/list/hiddenprints		//assoc ckey = realname/gloves/ckey
+	var/list/blood_DNA			//assoc dna = bloodtype
+	var/list/fibers				//assoc print = print
+
+/datum/component/forensics/InheritComponent(datum/component/forensics/F, original)
+	SANITIZE_LIST(fingerprints) |= F.fingerprints
+	SANITIZE_LIST(hiddenprints) |= F.hiddenprints
+	SANITIZE_LIST(blood_DNA) |= F.blood_DNA
+	SANITIZE_LIST(fibers) |= F.fibers
```
Comment (ninjanomnom) [inline code/datums/components/forensics.dm]: I hate how you do this component and it has enough wrong with it that I don't want to untangle the chains of logic and over sanitation. Get this in a more readable state.

Also the usecase of `SANITIZE_LIST` where it isn't a list already will result in nothing happening because this will be `list() |+ F.fibers` aka no asignment

File: code/datums/components/forensics.dm [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+	return ..()
+
+/datum/component/forensics/Initialize(new_fingerprints, new_hiddenprints, new_blood_DNA, new_fibers)
+	if(!isatom(parent))
+		. = COMPONENT_INCOMPATIBLE
+		CRASH("Forensics datum applied incorrectly to non-atom of type [parent.type]!")
+	add_fingerprint_list(new_fingerprints)
+	add_hiddenprint_list(new_hiddenprints)
+	add_blood_DNA(new_blood_DNA)
+	add_fiber_list(new_fibers)
```
Comment (ninjanomnom) [inline code/datums/components/forensics.dm]: they're null, stop it

`fingerprints = new_fingerprints`

File: code/datums/components/forensics.dm [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+	if(strength >= CLEAN_STRENGTH_FINGERPRINTS)
+		wipe_fingerprints()
+	if(strength >= CLEAN_STRENGTH_BLOOD)
+		wipe_blood_DNA()
+	if(strength >= CLEAN_STRENGTH_FIBERS)
+		wipe_fibers()
+
+/datum/component/forensics/proc/gc()	//unlikely to happen due to hiddenprints always persisting but you never know.
+	if(!length(hiddenprints) && !length(fingerprints) && !length(blood_DNA) && !length(fibers))
+		qdel(src)
```
Comment (ninjanomnom) [inline code/datums/components/forensics.dm]: just own it and say once a forensic component is added it never gets removed

File: code/game/objects/items/stacks/stack.dm [hunk 20]
```diff
@@ -326,9 +326,9 @@
 		. = ..()
 
 /obj/item/stack/proc/copy_evidences(obj/item/stack/from as obj)
```
Comment (ninjanomnom) [inline code/game/objects/items/stacks/stack.dm]: remove `as obj`

File: code/game/objects/structures/ladders.dm [hunk 21]
```diff
@@ -71,7 +71,6 @@
 
 /obj/structure/ladder/proc/travel(going_up, mob/user, is_ghost, obj/structure/ladder/ladder)
 	if(!is_ghost)
-		show_fluff_message(going_up, user)
```
Comment (ninjanomnom) [inline code/game/objects/structures/ladders.dm]: why was this removed?

File: code/modules/detectivework/detective_work.dm [hunk 22]
```diff
@@ -1,119 +1,135 @@
...
-			if(prob(20*item_multiplier) && !(fibertext in suit_fibers))
-				suit_fibers += "Material from a pair of [M.gloves.name]."
-	else if(M.gloves)
-		fibertext = "Material from a pair of [M.gloves.name]."
-		if(prob(20*item_multiplier) && !(fibertext in suit_fibers))
-			suit_fibers += "Material from a pair of [M.gloves.name]."
 
+/atom/proc/return_fingerprints()
+	GET_COMPONENT(D, /datum/component/forensics)
+	return D && length(D.fingerprints)? D.fingerprints : list()
```
Comment (ninjanomnom) [inline code/modules/detectivework/detective_work.dm]: too many mini procs that'll never be overridden with sanitation stacked on sanitation

**Review by ninjanomnom — DISMISSED** [review]

File: code/datums/components/forensics.dm [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+		return
+	LAZYINITLIST(blood_DNA)
+	for(var/i in dna)
+		blood_DNA[i] = dna[i]
+	return TRUE
+
+/datum/component/forensics/proc/signal_check_evidence()
+	. = NONE
+	if(length(fingerprints))
+		. |= FORENSICS_COMPONENT_HAS_FINGERPRINTS
```
Comment (ninjanomnom) [inline code/datums/components/forensics.dm]: @Cyberboss This sort of signal feels indicative of a bigger problem with the overall implementation but I'm unsure if it's a problem with @kevinz000 just doing it wrong or if it's a sign that a component is the wrong way to go instead of a plain datum on the atom path.

I'm leaning slightly more toward the latter

Comment (kevinz000) [reply code/datums/components/forensics.dm]: @Cyberboss

File: code/__DEFINES/components.dm [hunk 24]
```diff
@@ -77,3 +77,17 @@
 // /mob/living/carbon/human signals
 #define COMSIG_HUMAN_MELEE_UNARMED_ATTACK "human_melee_unarmed_attack"			//from mob/living/carbon/human/UnarmedAttack(): (atom/target)
 #define COMSIG_HUMAN_MELEE_UNARMED_ATTACKBY "human_melee_unarmed_attackby"		//from mob/living/carbon/human/UnarmedAttack(): (mob/living/carbon/human/attacker)
+
+//COMPONENT SPECIFIC SIGNALS
+// /datum/component/forensics
+#define COMSIG_FORENSICS_HAS_EVIDENCE "forensics_has_evidence"	//() - returns bitflags as shown below.
+	#define FORENSICS_COMPONENT_HAS_FINGERPRINTS 1
+	#define FORENSICS_COMPONENT_HAS_HIDDENPRINTS 2
+	#define FORENSICS_COMPONENT_HAS_FIBERS 4
```
Comment (Cyberboss) [inline code/__DEFINES/components.dm]: @ninjanomnom in response to your question, I'm going to go with the former. Signals shouldn't be written with the intent of having only one component listen for them. If that's the case, it's better to use the GetComponent family and call procs directly as it's much more efficient. Signals should be for events that several things would feasibly want to listen for. What other than forensics would want to listen for COMSIG_FORENSICS_HAS_EVIDENCE?

File: code/modules/clothing/gloves/_gloves.dm [hunk 25]
```diff
@@ -11,13 +11,21 @@
...
+		return
+	transfer_blood = 0
 
 /obj/item/clothing/gloves/worn_overlays(isinhands = FALSE)
 	. = list()
 	if(!isinhands)
 		if(damaged_clothes)
 			. += mutable_appearance('icons/effects/item_damage.dmi', "damagedgloves")
-		if(blood_DNA)
+		if(SendSignal(COMSIG_FORENSICS_HAS_EVIDENCE) & FORENSICS_COMPONENT_HAS_BLOOD_DNA)
```
Comment (Cyberboss) [inline code/modules/clothing/gloves/_gloves.dm]: All these send signals should just be getcomponent followed by querying the component for whatever it wants

**Review by Cyberboss — DISMISSED** [review]

File: code/__DEFINES/components.dm (hunk 24 above)
Comment (ninjanomnom) [reply code/__DEFINES/components.dm]: That there are enough places to want something like this though is the problem. When it's being accessed enough that you want to standardize the access, it starts feeling like it'd be better as a datum.

File: code/game/atoms.dm [hunk 26]
```diff
@@ -330,100 +322,28 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
...
-/atom/proc/transfer_blood_dna(list/blood_dna)
-	if(!blood_DNA)
-		blood_DNA = list()
-	var/old_length = blood_DNA.len
-	blood_DNA |= blood_dna
-	if(blood_DNA.len > old_length)
-		return 1//some new blood DNA was added
-
+		return FALSE
+	var/old_length = length(return_blood_DNA())
```
Comment (optimumtact) [inline code/game/atoms.dm]: you should really write a get_length proc or something

File: code/game/gamemodes/devil/true_devil/_true_devil.dm [hunk 27]
```diff
@@ -67,7 +67,8 @@
 	//Left hand items
 	for(var/obj/item/I in held_items)
 		if(!(I.flags_1 & ABSTRACT_1))
-			if(I.blood_DNA)
+			GET_COMPONENT_FROM(FR, /datum/component/forensics, I)
```
Comment (optimumtact) [inline code/game/gamemodes/devil/true_devil/_true_devil.dm]: you do this exact two pattern line so many times you should make it a macro or something

**Comment by kevinz000** [comment]
kk

**Comment by MoreRobustThanYou** [comment]
DID YOU JUST NOT SQUASH IT? 😮

File: code/game/machinery/washing_machine.dm [hunk 28]
```diff
@@ -12,6 +12,10 @@
 	var/obj/item/color_source
 	var/max_wash_capacity = 5
 
+/obj/machinery/washing_machine/Initialize()
+	. = ..()
+	AddComponent(/datum/component/redirect, list(COMSIG_COMPONENT_CLEAN_ACT), CALLBACK(src, .proc/clean_blood))
```
Comment (ninjanomnom) [inline code/game/machinery/washing_machine.dm]: ComponentInitialize

File: code/modules/clothing/gloves/_gloves.dm [hunk 29]
```diff
@@ -11,13 +11,21 @@
 	strip_delay = 20
 	equip_delay_other = 40
 
+/obj/item/clothing/gloves/Initialize()
+	. = ..()
+	AddComponent(/datum/component/redirect, list(COMSIG_COMPONENT_CLEAN_ACT), CALLBACK(src, .proc/clean_blood))
```
Comment (ninjanomnom) [inline code/modules/clothing/gloves/_gloves.dm]: ComponentInitialize

File: code/modules/clothing/shoes/_shoes.dm [hunk 30]
```diff
@@ -15,12 +15,16 @@
 	var/offset = 0
 	var/equipped_before_drop = FALSE
 
+/obj/item/clothing/shoes/Initialize()
+	. = ..()
+	AddComponent(/datum/component/redirect, list(COMSIG_COMPONENT_CLEAN_ACT), CALLBACK(src, .proc/clean_blood))
```
Comment (ninjanomnom) [inline code/modules/clothing/shoes/_shoes.dm]: ComponentInitialize

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/game/atoms.dm [hunk 31]
```diff
@@ -239,7 +235,8 @@
 /atom/proc/examine(mob/user)
 	//This reformat names to get a/an properly working on item descriptions when they are bloody
 	var/f_name = "\a [src]."
-	if(src.blood_DNA && !istype(src, /obj/effect/decal))
+	GET_COMPONENT(forensics, /datum/component/forensics)
+	if(forensics && length(forensics.blood_DNA) && !istype(src, /obj/effect/decal))
```
[the remark this answers is not in the document]
Comment (ninjanomnom) [reply code/game/atoms.dm]: To expand on what i said in irc:
get f_name like
```
var/list/f_name_list = list("\a", "[src]")
if(SendSignal(COMSIG_ATOM_F_NAME, f_name_list) & F_NAME_CHANGED)
  f_name = f_name_list.Join

<inside the component proc>

f_name_list.len = 1
f_name_list[1] = "some <span class='danger'>blood-stained</span> [parent]!"
return F_NAME_CHANGED
```

File: code/game/atoms.dm [hunk 32]
```diff
@@ -228,22 +228,22 @@
...
 	else if(src in container)
-		return 1
+		return TRUE
+	return FALSE
+
+/atom/proc/get_examine_name(mob/user)
+	var/list/inject_before = list()
+	var/list/inject_after = list()
+	SendSignal(COMSIG_ATOM_GET_EXAMINE_NAME, user, inject_before, inject_after)
+	. = "[gender == PLURAL? "some" : "a"] [length(inject_before)? "[english_list(inject_before, "")] " : ""][src][length(inject_after)? " [english_list(inject_after, "")]" : ""]"
```
Comment (ninjanomnom) [inline code/game/atoms.dm]: You don't need these before/after lists and what if a component wants to modify the name itself and hide it?
```
. = "\a [src]"
var/list/f_name_list = list("\a", " ", "[src]")
if(SendSignal(COMSIG_ATOM_F_NAME, f_name_list) & F_NAME_CHANGED)
  . = f_name_list.Join("")
```

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/detectivework/detective_work.dm [hunk 33]
```diff
@@ -1,119 +1,111 @@
...
+/obj/add_blood_DNA(list/dna)
+	. = ..()
+	if(length(dna))
+		. = AddComponent(/datum/component/forensics, null, null, dna)
+
+/obj/item/add_blood_DNA(list/blood_dna, list/datum/disease/diseases)
+	. = ..()
+	if(!.)
+		return FALSE
+	GET_COMPONENT(FR, /datum/component/forensics)
```
Comment (ninjanomnom) [inline code/modules/detectivework/detective_work.dm]: `.` is the component you just added

File: code/modules/detectivework/detective_work.dm [hunk 34]
```diff
@@ -1,119 +1,111 @@
...
+	if(FR && length(FR.blood_DNA))
+		AddComponent(/datum/component/decal/blood)
```
Comment (ninjanomnom) [inline code/modules/detectivework/detective_work.dm]: the forensics components should handle adding/removing the blood decal

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by optimumtact** [comment]
MoreRobustThanYou's datum forensics but it's a component this time.
Fully refactors forensics, including fingerprints, hidden/admin fingerprints, suit/clothing/whatever fibers, and blood DNA residue to /datum/component/forensics.
Adds support for manually adding fingerprints.
:cl:
code: Forensics is now a datum component.
balance: NPC humans will now start leaving fingerprints on things they touch!
/:cl:
